fix: image slices are displayed in reverse order, if we switch from common layout to 2D MPR

arul-trenser opened the pull request:
…like ImagePositionPatient/ImageOrientationPatient is available and sorted stack images by instance number when necessary metadata is not present

<!-- Do Not Delete This! pr_template -->
<!-- Please read our Rules of Conduct: https://github.com/OHIF/Viewers/blob/master/CODE_OF_CONDUCT.md -->
<!-- 🕮 Read our guide about our Contributing Guide here https://docs.ohif.org/development/contributing -->
<!-- :hand: Thank you for starting this amazing contribution! -->

<!--
⚠️⚠️ Please make sure the checklist section below is complete before submitting your PR.
To complete the checklist, add an 'x' to each item: [] -> [x]
(PRs that do not have all the checkboxes marked will not be approved)
-->

### Context

- Fixes #5021 
- This PR has been incorporated by [FlyWheel.io](https://flywheel.io/).

<!--
Provide a clear explanation of the reasoning behind this change, such as:
- A link to the issue being addressed, using the format "Fixes #ISSUE_NUMBER"
- An image showing the issue or problem being addressed (if not already in the issue)
- Error logs or callStacks to help with the understanding of the problem (if not already in the issue)
-->

### Changes & Results

This issue is fixed by sorting slices according to spatial position when necessary metadata like `ImagePositionPatient`/`ImageOrientationPatient` is available and sorted stack images by instance number when necessary metadata is not present.


https://github.com/user-attachments/assets/dd249c2a-c42b-402b-8af4-1c8bc7a7cb09


<!--
List all the changes that have been done, such as:
- Add new components
- Remove old components
- Update dependencies

What are the effects of this change?
- Before vs After
- Screenshots / GIFs / Videos
-->

### Testing

<!--
Describe how we can test your changes.
- open a URL
- visit a page
- click on a button
- etc.
-->

### Checklist

#### PR

<!--
https://semantic-release.gitbook.io/semantic-release/#how-does-it-work

Examples:
Please note the letter casing in the provided examples (upper or lower).

- feat(MeasurementService): add ...
- fix(Toolbar): fix ...
- docs(Readme): update ...
- style(Whitespace): fix ...
- refactor(ExtensionManager): ...
- test(HangingProtocol): Add test ...
- chore(git): update ...
- perf(VolumeLoader): ...

You don't need to have each commit within the Pull Request follow the rule,
but the PR title must comply with it, as it will be used as the commit message
after the commits are squashed.
-->

- [x] My Pull Request title is descriptive, accurate and follows the
  semantic-release format and guidelines.

#### Code

- [x] My code has been well-documented (function documentation, inline comments,
  etc.)

#### Public Documentation Updates

<!-- https://docs.ohif.org/ -->

- [x] The documentation page has been updated as necessary for any public API
  additions or removals.

#### Tested Environment

- [x] OS: <!--[e.g. Windows 10, macOS 10.15.4]-->  Windows 11
- [x] Node version: <!--[e.g. 18.16.1]--> 18.19.0
- [x] Browser: Chrome 131.0.6778.265
  <!--[e.g. Chrome 83.0.4103.116, Firefox 77.0.1, Safari 13.1.1]-->

<!-- prettier-ignore-start -->
[blog]: https://circleci.com/blog/triggering-trusted-ci-jobs-on-untrusted-forks/
[script]: https://github.com/jklukas/git-push-fork-to-upstream-branch
<!-- prettier-ignore-end -->

wayfarer3130 commented on extensions/default/src/utils/validations/areAllImagePositionsEqual.ts:
```diff
@@ -1,7 +1,7 @@
 import toNumber from '@ohif/core/src/utils/toNumber';
 import { _getPerpendicularDistance } from '@ohif/core/src/utils/isDisplaySetReconstructable';
-import calculateScanAxisNormal from '../calculateScanAxisNormal';
+import calculateScanAxisNormal from '@ohif/core/src/utils/calculateScanAxisNormal';
```
You can't access src directory from @ohif/core - I know it mostly builds, but not consistently.  You will have to export it from the utils module in @ohif/core and then import it from @ohif/core and then extract from utils.

wayfarer3130 commented on platform/core/src/classes/ImageSet.ts:
```diff
@@ -79,63 +75,96 @@ class ImageSet {
+  sort(sortingCallback: (a: Image, b: Image) => number): Image[] {
+    try {
+      return this.sortByImagePositionPatient();
+    } catch (error) {
```
Catching an error for a normal flow isn't the right way to handle this - what you should do is something like:
const sortImagePositionCompare = createSortByImagePositionPatient(images);
this.images.sort(sortImagePositionCompare || defaultImageSort)

wayfarer3130 commented on platform/core/src/classes/ImageSet.ts:
```diff
@@ -79,63 +75,96 @@ class ImageSet {
+
+    // Compute distances from each image to the reference image
+    const distanceInstancePairs = images.map(image => {
+      const imagePositionPatient = image.ImagePositionPatient;
```
Please use the 
const deltaVector = vec3.create();
...
vec3.dot(scanAxisNormal, vec3.subtract(deltaVector,imagePositionPatient, referenceImagePositionPatient)) to calculate distances rather than re-implementing it dot product using reduce.  It makes it clearer what is going on.

wayfarer3130 commented on platform/core/src/classes/ImageSet.ts:
```diff
@@ -1,15 +1,11 @@
-      ImageOrientationPatient: number[];
-    };
-  };
+  ImagePositionPatient?: string;
```
This shouldn't be a string ever - it is a Point3 if defined, same with ImageOrientationPatient.

wayfarer3130 commented on platform/core/src/classes/ImageSet.ts:
```diff
@@ -79,63 +75,96 @@ class ImageSet {
+   *
+   * @returns images - reference to images after sorting
+   */
+  sortByInstanceNumber(): Image[] {
```
This should use the public sort instances in the sortStudy utility - you will probably have to export the instancesSortCriteria, although the thumbnail PR adds that already, and will probably merge before your PR does.

arul-trenser replied:
@wayfarer3130  Modified as suggested.

wayfarer3130 commented on platform/core/src/classes/ImageSet.ts:
```diff
@@ -79,63 +75,96 @@ class ImageSet {
+   *
+   * @returns images - reference to images after sorting
+   */
+  sortByImagePositionPatient(): Image[] {
```
With the remaining fixes for making this a comparator, the createImagePositionPatientCompare should get moved into sortStudy so that it can be re-used consistently.

arul-trenser commented:
@wayfarer3130  I have made the updates to the PR based on your review comments. Could you please take a look when you have a moment?

Thanks!

wayfarer3130 commented on platform/core/src/utils/sortStudy.ts:
```diff
@@ -128,11 +130,60 @@ export default function sortStudy(
+ *
+ * @returns images - reference to images after sorting
+ */
+const sortStudyByImagePositionPatient = (images, fallbackSort) => {
```
Can you pass a comparator in rather than a full sort function?

wayfarer3130 reviewed (commented):
Almost there - the only critical fix is to change the final sort order.  Using indexOf is an O(n) function, so doing it the way you have would make the sort O(n*n*log n) which is too slow.

arul-trenser commented:
@wayfarer3130  I have updated as per the suggestions. Please review at your convenience. 
Thanks!

sedghi reviewed (changes requested):
1) We need migration guide in the https://docs.ohif.org/migration-guide/3p10-to-3p11/ explains our sorting has changed, and how they can switch back to the old way (we should provide away to say add this configuration and it will use the old way

Otherwise looks great

arul-trenser commented:
@sedghi  I have added a customization to switch back to the old way also updated the documentation.
Could you please review it at you convenience?

wayfarer3130 commented on platform/core/src/classes/ImageSet.ts:
```diff
@@ -1,15 +1,16 @@
+} from '../utils/sortStudy';
 
 type Attributes = Record<string, unknown>;
 type Image = {
```
export type Image here so it is accessible.

arul-trenser replied:
changed as per your suggestion

wayfarer3130 reviewed (approved):
Thinking about where the customization goes, I think the current behaviour is good for now, but eventually having more control over the comparison, eg to have an init method or a factor method available to provide context would be useful, and then using straight comparators.  That is something that can be added later and I think this is a good change as it is now.

IbrahimCSAE commented:
The changes in the PR broke some of the sorting on my data, I will see if I can get permission to post an anonymized version, but after I reverted the changes from this PR, the sorting worked again as expected

arul-trenser commented:
> Can you look at the failing comparisons?

@wayfarer3130 The Playwright test `jumpToMeasurementMPR` initialDraw is failing, likely because the screenshot is being captured before the image has fully loaded. This issue appears to be recurring in other pull requests as well.

arul-trenser commented on platform/core/src/classes/ImageSet.ts:
```diff
@@ -79,63 +80,46 @@ class ImageSet {
+   *
+   * @returns images - reference to images after sorting
+   */
+  sort(customizationService): Image[] {
```
@sedghi Could you please share your thoughts on this?

Thanks!

jbocce commented:
@arul-trenser , since I have been working on the e2e tests I figured I would have a look for you. There are two failing tests:

```
  2 failed
    [chromium] › tests/JumpToMeasurementMPR.spec.ts:10:5 › should hydrate in MPR correctly ─────────
    [chromium] › tests/MPRThenRTOverlayNoHydration.spec.ts:10:5 › should launch MPR with unhydrated RTSTRUCT chosen from the data overlay menu 
```

The `MPRThenRTOverlayNoHydration.spec.ts` you can ignore. I will be improving those tests like this one where there are pop ups soon.

However `JumpToMeasurementMPR.spec.ts` is something you should look into - be sure to look at the screen captures of the actual and expected to help you out. I can see the problem reported by testing manually as well. One of the first things that the test does is navigate to image 21. However, this PR seems to have changed what image 21 actually is. So either change the screenshot (be sure to add a 5 second wait prior to taking the screen shot) OR fix the problem.

sedghi commented:
We're wrapping up a release, so if you'd like to include this, please address all comments by the end of the week.

wayfarer3130 commented:
> > @jbocce Previously, the study appeared reversed in the common layout, although it was correct in the MPR layout. This PR addresses that reversal. As a result, the slice order in the common layout will now be different, which is the expected behavior and resolves the underlying issue. This change caused a mismatch with existing screenshots, leading to test failures, so I've updated them accordingly.
> > However, I've observed that the display set order in the left side panel is inconsistent when running Playwright tests. This issue is difficult to reproduce manually, occurring only about once in 50 attempts, and typically during the initial un-cached load. It's also not consistently reproducible even under those conditions (in master). However, the order becomes correct if we reload the page or navigate back to the study list and re-select the study. This order mismatch is causing a test failure in the master branch, which I've also confirmed locally.
> > cc: @sedghi
> 
> Yes I understand. I think it might be best to simply close the left panel prior to taking screenshots and open as necessary with `await page.getByTestId('side-panel-header-left').click();`.

Someone removed the sort order from the display sets on startup.  That has to get added back in again.  Adding it in will make the display consistent in the left hand panel.  That should be a one line change.

sedghi reviewed (changes requested):
please add a migration guides too, to tell people about the new sorting, and how they can switch back to the old sorting (if you have not this fallback we should)

sedghi commented:
can you check cypress why it fails and fix it

IbrahimCSAE commented:
FYI. This has broken the sort on some series, they became completely out of order. I can't share the data but the only information I can share is that it was a sagittal series with an axial cover slice.

Its sorted correctly on viewer.ohif.org, but on viewer-dev.ohif.org is completely messed up @sedghi

arul-trenser commented:
> FYI. This has broken the sort on some series, they became completely out of order. I can't share the data but the only information I can share is that it was a sagittal series with an axial cover slice.
> 
> Its sorted correctly on viewer.ohif.org, but on viewer-dev.ohif.org is completely messed up @sedghi

The issue mentioned above has been addressed. I've implemented a fix in PR #5223.
 
@IbrahimCSAE  Could you please confirm if this resolves your data sorting problem?

IbrahimCSAE commented:
Its resolved now, Thanks!

arul-trenser commented:
@sedghi Could you please review #5223 and approve tha PR

> > FYI. This has broken the sort on some series, they became completely out of order. I can't share the data but the only information I can share is that it was a sagittal series with an axial cover slice.
> > Its sorted correctly on viewer.ohif.org, but on viewer-dev.ohif.org is completely messed up @sedghi
> 
> The issue mentioned above has been addressed. I've implemented a fix in PR #5223.
> 
> @IbrahimCSAE Could you please confirm if this resolves your data sorting problem?

@sedghi Could you please review the PR #5223, I've fixed the above mentioned issue.
